numpy/core/src/multiarray/multiarraymodule.c

```diff
@@ -813,121 +813,72 @@ new_array_for_sum(PyArrayObject *ap1, PyArrayObject *ap2, PyArrayObject* out,
 NPY_NO_EXPORT PyObject *
 PyArray_InnerProduct(PyObject *op1, PyObject *op2)
 {
-    PyArrayObject *ap1, *ap2, *ret = NULL;
-    PyArrayIterObject *it1, *it2;
-    npy_intp i, j, l;
-    int typenum, nd, axis;
-    npy_intp is1, is2, os;
-    char *op;
-    npy_intp dimensions[NPY_MAXDIMS];
-    PyArray_DotFunc *dot;
-    PyArray_Descr *typec;
-    NPY_BEGIN_THREADS_DEF;
+    PyArrayObject *ap1 = NULL;
+    PyArrayObject *ap2 = NULL;
+    int typenum;
+    PyArray_Descr *typec = NULL;
+    PyObject* ap2t = NULL;
+    npy_intp dims[NPY_MAXDIMS];
+    PyArray_Dims newaxes = {dims, 0};
+    int i;
+    PyObject* ret = NULL;
 
     typenum = PyArray_ObjectType(op1, 0);
     typenum = PyArray_ObjectType(op2, typenum);
-
     typec = PyArray_DescrFromType(typenum);
     if (typec == NULL) {
-        return NULL;
+        PyErr_SetString(PyExc_ValueError, "Cannot find a common data type.");
```

Comment:
So, I tried the following with and without are changes and got the same result in either case. In short, it never hits our error message. Also, it gives two different error messages depending on the order.

    >>> import numpy
    >>> import numpy as np
    >>> np.inner(1., np.array((1,1), dtype='i,i'))
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    TypeError: invalid type promotion
    >>> np.inner(np.array((1,1), dtype='i,i'), 1.)
    Traceback (most recent call last):
      File "<stdin>", line 1, in <module>
    TypeError: Cannot cast array data from dtype([('f0', '<i4'), ('f1', '<i4')]) to dtype('float64') according to the rule 'safe'